Ad9208 support

stefpopa opened the pull request:
This PR adds the ad9208 driver, the ad9208 api and the project files.

mirceacaprioru commented:
Should we exclude the api from our review ?

stefpopa commented:
> Should we exclude the api from our review ?

I think that reviewing the API is pointless

dbogdan commented:
The ADC driver must go to drivers/adc folder.

dbogdan commented on drivers/ad9208/ad9208.h:
```diff
@@ -0,0 +1,157 @@
+	jesd_param_t *jesd_param;
+	uint32_t jesd_subclass;
+} ad9208_init_param;
+
```
/******************************************************************************/
/************************ Functions Declarations ******************************/
/******************************************************************************/

/*! Initialize the device. */

dbogdan commented on drivers/ad9208/ad9208.h:
```diff
@@ -0,0 +1,157 @@
+*******************************************************************************/
+#ifndef __AD9208_H__
+#define __AD9208_H__
+
```
/******************************************************************************/
/***************************** Include Files **********************************/
/******************************************************************************/

mirceacaprioru commented on drivers/ad9208/ad9208.h:
```diff
@@ -0,0 +1,157 @@
+#include <stdbool.h>
+#include "ad9208_api.h"
+#include "ad9208_reg.h"
+#include "platform_drivers.h"
```
aren't we moving to specific headers and implementations for platform drivers ? (eg. spi, i2c )

stefpopa replied:
I think that the xilinx_platform still uses platform_drivers: https://github.com/analogdevicesinc/no-OS/tree/master/projects/drivers/xilinx_platform

mirceacaprioru commented on drivers/ad9208/ad9208.c:
```diff
@@ -0,0 +1,514 @@
+	uint8_t * buffer = (uint8_t *) malloc(len);
+	int32_t ret;
+
+	buffer[0] = wbuf[0];
```
looks like memcpy

mirceacaprioru commented on drivers/ad9208/ad9208.c:
```diff
@@ -0,0 +1,514 @@
+	ad9208_handle_t *ad9208_h = st->adc_h;
+	int ret;
+
+ 	ret = ad9208_adc_set_channel_select(ad9208_h, BIT(chan & 1));
```
does this work only for channels 1, 3, 5 ? 
BIT(chan & 1)

stefpopa replied:
I think it's implemented like that because there are only 2 channels that you should be able to set: channel 0 and channel 1

commodo commented:
codacy has 1 complaint
```
Clarify calculation precedence for '&' and '?'.
        printf("AD9208 PLL %s\n", pll_stat & AD9208_JESD_PLL_LOCK_STAT ?
```

commodo commented:
> > Should we exclude the api from our review ?
> 
> I think that reviewing the API is pointless

maybe include an astyle ignore thingi for the API, so that Travis-CI won't complain
@amiclaus can help with that

mirceacaprioru commented on drivers/ad9208/ad9208.c:
```diff
@@ -0,0 +1,514 @@
+ 	ret |= spi_remove(device->spi_desc);
+
+ 	free(device->st);
+
```
free(st->jesd_param), free(st->adc_h) ?

commodo commented on drivers/ad9208/ad9208.c:
```diff
@@ -0,0 +1,514 @@
+
+ 	return 0;
+
+error_7:
```
as a coding-pattern, these `goto error_` groups look hard to maintain/change;
that is: they require a bit of brain power to follow and make sure they are not leaking;
for code that is final, and is sure to not change, it's probably fine; their speed/run-time efficiency remains uncontested;

not sure how often this code would change to make things sure we don't introduce leaks/errors at a later change;

the alternative here, is to create a single `error` label and do:
```
    if (st->adc_h)
        free(st->adc_h);
    if (st->jesd_param)
        free(st->jesd_param);
    ....
```

of course, you'd need to make sure that `st` is initialized with zeros, so that non-NULL is free'd properly;
and the null-checks are required here, as not all implementations of `free()` are guaranteed to ignore `free(NULL)`

stefpopa commented:
changes in v2:
* added doxygen comments for the functions
* moved the driver in `drivers/adc/ad9208` folder
* fixed codacy complaint
* used only one error label in `ad9208_initialize()`

cristiantpop commented on projects/ad9208/main.c:
```diff
@@ -0,0 +1,480 @@
+		RX_DMA_BASEADDR,
+		DMA_DEV_TO_MEM,
+		0,
+	};
```
Probably we should use explicit names, for structure elements, when initializing :
.name  = "rx_dmac"

cristiantpop commented on drivers/adc/ad9208/ad9208.c:
```diff
@@ -0,0 +1,517 @@
+	adi_chip_id_t chip_id;
+	uint64_t sample_rate, lane_rate_kbps;
+	uint8_t dcm, pll_stat;
+	int timeout, i, ret;
```
Use stdint types.
int timeout, i, ret;

cristiantpop commented on projects/ad9208/main.c:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	status = axi_dmac_transfer(rx_dmac,
+				   DDR_MEM_BASEADDR + 0x800000, 16384 * 8);
```
Probably we should use number of channels here: 16384 * 8

stefpopa commented:
changes in v3:
* fixed astyle warnings
* used stdint types
* used explicit names, for structure elements, when initializing

mirceacaprioru commented:
you still have some small styling issues

stefpopa commented:
> you still have some small styling issues
I don't necessarily agree with this one
```
 	ret = ad9208_set_input_clk_duty_cycle_stabilizer(ad9208_h,
-							 st->duty_cycle_stabilizer_en);
+			st->duty_cycle_stabilizer_en);
```

The rest seem to be in the API or different files that I haven't touched for this PR.
